# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/vr/VRManager.cpp
# Commit: ec6a64b639a8
# Full Hash: ec6a64b639a82e957376c09a71aef7ce62b3a5aa
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-12-03 09:47:26
# Regressor Bug: 1657404
# File Overlap Count: 1
# Description:
#   Bug 1657404 - Implement a strongly-typed, service-dependent gamepad handle r=handyman,aklotz
#   
#   Currently, the gamepad code uses a uint32_t as a handle and does some trickery
#   with it by trying to create a unique ID and adding an offset to it for VR code.
#   
# ==============================================================================

diff -r 6792f28b99bd -r ec6a64b639a8 gfx/vr/VRManager.cpp
--- a/gfx/vr/VRManager.cpp	Wed Dec 02 22:36:35 2020 +0000
+++ b/gfx/vr/VRManager.cpp	Wed Dec 02 23:06:05 2020 +0000
@@ -53,6 +53,8 @@
 using namespace mozilla::layers;
 using namespace mozilla::gl;
 
+using mozilla::dom::GamepadHandle;
+
 namespace mozilla::gfx {
 
 /**
@@ -864,8 +866,9 @@
   PushState();
 }
 
-void VRManager::VibrateHaptic(uint32_t aControllerIdx, uint32_t aHapticIndex,
-                              double aIntensity, double aDuration,
+void VRManager::VibrateHaptic(GamepadHandle aGamepadHandle,
+                              uint32_t aHapticIndex, double aIntensity,
+                              double aDuration,
                               const VRManagerPromise& aPromise)
 
 {
@@ -877,7 +880,7 @@
   // understood by VRDisplayExternal.
   uint32_t controllerBaseIndex =
       kVRControllerMaxCount * mDisplayInfo.mDisplayID;
-  uint32_t controllerIndex = aControllerIdx - controllerBaseIndex;
+  uint32_t controllerIndex = aGamepadHandle.GetValue() - controllerBaseIndex;
 
   TimeStamp now = TimeStamp::Now();
   size_t bestSlotIndex = 0;
@@ -934,7 +937,7 @@
   PushState();
 }
 
-void VRManager::StopVibrateHaptic(uint32_t aControllerIdx) {
+void VRManager::StopVibrateHaptic(GamepadHandle aGamepadHandle) {
   if (mState != VRManagerState::Active) {
     return;
   }
@@ -943,7 +946,7 @@
   // understood by VRDisplayExternal.
   uint32_t controllerBaseIndex =
       kVRControllerMaxCount * mDisplayInfo.mDisplayID;
-  uint32_t controllerIndex = aControllerIdx - controllerBaseIndex;
+  uint32_t controllerIndex = aGamepadHandle.GetValue() - controllerBaseIndex;
 
   for (size_t i = 0; i < mozilla::ArrayLength(mBrowserState.hapticState); i++) {
     VRHapticState& state = mBrowserState.hapticState[i];